### ISSUE TITLE
Incorrect discriminated union narrowing following exhaustive if/else

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#9246]
```ts
type Shape = { kind: "Circle"; radius: number }
           | { kind: "Rectangle"; width: number; height: number }


function area(s: Shape) {
    if (s.kind === "Circle") {
        return s.radius ** 2;
    }
    else if (s.kind === "Rectangle") {
        return s.width * s.height;
    }
    return assertNever(s);
}

function assertNever(x: never): never {
    return x;
}
```

**Expected:** No error
**Actual:** 

```text
test.ts(12,24): error TS2345: Argument of type '{ kind: "Circle"; radius: number; }' is not assignable to parameter of type 'never'.
```

**Comment by ahejlsberg** [comment]
This is by design. It is an effect of #8548.

**Comment by ahejlsberg** [comment]
This one will be tricky to fix. We basically have two conflicting goals: Sometimes we want to use control flow analysis to prove exhaustiveness, at other times we want a convenient way to defeat the exhaustiveness (e.g. when writing defensive code or when control flow analysis can't see the entire picture).

Part of the problem is that we can't always see the "connectedness" of multiple type guards. In a `switch` statement it is pretty obvious that all of the cases are part of the same test, and it is therefore reasonable to allow an exhaustive check to produce the `never` type. But with multiple `if` statements it gets harder because they could be written as a single connected `if ... else if ... else if ... else` or as multiple separate `if`s, possibly with intervening code. Logically it is all the same, but the intent may have been different.
